src/services/StructuralMetadataUtils.test.js

```diff
@@ -0,0 +1,593 @@
+      expect(value).toHaveLength(expected.length);
+      expect(value).toContainEqual({
+        type: 'div',
+        label: 'Sub-Segment 2.1',
+        id: '123a-456b-789c-6d'
+      });
+    });
+    test('no wrapping timespans after', () => {
+      const wrapperSpans = {
+        before: {
+          type: 'span',
+          label: 'Segment 2.1',
+          id: '123a-456b-789c-8d',
+          begin: '00:09:03.24',
+          end: '00:15:00.00'
+        },
+        after: null
+      };
+      const expected = [
+        {
+          type: 'div',
+          label: 'Sub-Segment 2.1',
+          id: '123a-456b-789c-6d'
+        }
+      ];
+      const value = smu.getValidHeadings(wrapperSpans, testData);
+      expect(value).toHaveLength(expected.length);
+      expect(value).toContainEqual({
+        type: 'div',
+        label: 'Sub-Segment 2.1',
+        id: '123a-456b-789c-6d'
+      });
+    });
+  });
+
+  describe('finds valid headings when either timespans before or after is null', () => {
+    test('wrapping timespan before in null', () => {
+      const wrapperSpans = {
+        before: null,
+        after: {
+          type: 'span',
+          label: 'Act 1',
+          id: '123a-456b-789c-3d',
+          begin: '00:10:00.00',
+          end: '00:15:00.00'
+        }
+      };
+      const value = smu.getValidHeadingForEmptySpans(
+        wrapperSpans,
+        testEmptyHeaderBefore
+      );
+      expect(value).toEqual([
+        { type: 'div', label: 'Scene 1', id: '123a-456b-789c-1d' }
+      ]);
+    });
+    test('wrapping timespan after is null', () => {
+      const wrapperSpans = {
+        before: {
+          type: 'span',
+          label: 'Act 1',
+          id: '123a-456b-789c-2d',
+          begin: '00:00:00.00',
+          end: '00:09:00.00'
+        },
+        after: null
+      };
+      const value = smu.getValidHeadingForEmptySpans(
+        wrapperSpans,
+        testEmptyHeaderAfter
+      );
+      expect(value).toEqual([
+        { type: 'div', label: 'Scene 2', id: '123a-456b-789c-3d' }
+      ]);
+    });
+  });
+
+  describe('validates order of begin and end times', () => {
+    test('begin time < end time', () => {
+      const begin = '00:00:10.30';
+      const end = '00:10:00.30';
+      const vlaue = smu.validateBeforeEndTimeOrder(begin, end);
+      expect(vlaue).toBeTruthy();
+    });
+    test('begin time === end time', () => {
+      const begin = '00:00:10.30';
+      const end = '00:00:10.30';
+      const vlaue = smu.validateBeforeEndTimeOrder(begin, end);
+      expect(vlaue).toBeFalsy();
+    });
+    test('begin time > end time', () => {
+      const begin = '00:10:00.30';
+      const end = '00:00:10.30';
+      const vlaue = smu.validateBeforeEndTimeOrder(begin, end);
+      expect(vlaue).toBeFalsy();
+    });
+  });
+
+  describe('tests deleting list item', () => {
+    test('deletes a timespan from metadata structure', () => {
```

Comment:
Another common pattern you'll see in JS unit testing is using a `beforeEach()` or `beforeAll()` function in a group of tests, which in this case could grab a fresh reference to `testData` before running each test.  It's a way to not mutate an object when running multiple assertions against some mock object.  Or to re setup your data before running multiple tests against it.

Reply:
@adamjarling Oh yeah, I'll try that. Thanks :+1:

Reply:
@adamjarling I used `beforeEach()` to initialize the `testData` array, but still gets the same issue of changes in the original array when testing `deleteListItem()`. 
I think I should keep this for now, and do something else.